File: scala/src/main/scala/Vector.scala
package dependent

import scala.language.implicitConversions

import nat._

object Vector {
  trait Vec[+A, N <: Nat]

  case object Nil extends Vec[Nothing, Z]

  case class Cons[+A, N <: Nat](h: A, t: Vec[A, N]) extends Vec[A, Succ[N]]

  implicit def narrow[A, N <: Nat](v: Vec[A, Succ[N]]): Cons[A, N] = v match {
    case vc @ Cons(_, _) => vc
  }

  trait Rep[A, N <: Nat] {
    def apply(a: A): Vec[A, N]
  }

  implicit def zRep[A] = new Rep[A, Z] {
    def apply(a: A) = Nil
  }

  implicit def nRep[A, N <: Nat](implicit nat: Rep[A, N]) =
    new Rep[A, Succ[N]] {
      def apply(a: A) = Cons(a, rep[A, N](a))
    }

  def rep[A, N <: Nat](a: A)
    (implicit r: Rep[A, N]): Vec[A, N] = r(a)

  trait Concat[A, XS, YS] {
    type MN <: Nat
    type Out = Vec[A, MN]
    def apply(xs: XS, ys: YS): Out
  }

  implicit def nilConcat[A, N <: Nat] =
    new Concat[A, Vec[A, Z], Vec[A, N]] {
      type MN = N
      def apply(xs: Vec[A, Z], ys: Vec[A, N]): Out = ys
    }

  implicit def conConcat[A, M <: Nat, N <: Nat]
    (implicit tail: Concat[A, Vec[A, M], Vec[A, N]]) =
    new Concat[A, Vec[A, Succ[M]], Vec[A, N]] {
      type MN = Succ[tail.MN]
      def apply(xs: Vec[A, Succ[M]], ys: Vec[A, N]): Out =
        Cons(xs.h, app(xs.t, ys))
    }

  def app[A, N <: Nat, M <: Nat](xs: Vec[A, N], ys: Vec[A, M])
    (implicit concat: Concat[A, Vec[A, N], Vec[A, M]]): concat.Out =
    concat(xs, ys)
}

File: hs/src/Vector.hs
{-# LANGUAGE GADTs, DataKinds, KindSignatures, TypeFamilies #-}

module Vector where

data Nat = Z | S Nat

data Vec :: * -> Nat -> * where
  Nil :: Vec a Z
  Cons :: a -> Vec a n -> Vec a (S n)

data SNat n where
  SZ :: SNat Z
  SS :: SNat n -> SNat (S n)

rep :: SNat n -> a -> Vec a n
rep SZ a = Nil
rep (SS n) a = Cons a (rep n a)

type family Plus(n :: Nat)(m :: Nat) :: Nat where
  Plus Z m = m
  Plus (S n) m = S (Plus n m)

sPlus :: SNat n -> SNat m -> SNat (Plus n m)
sPlus SZ m = m
sPlus (SS n) m = SS (sPlus n m)

app :: Vec a n -> Vec a m -> Vec a (Plus n m)
app Nil l = l
app (Cons h t) l = Cons h (app t l)
